From: sunshineinabox <aqemail@gmail.com>
Date: Mon, 27 Jul 2026 20:40:08 -0700
Subject: [PATCH] ASoC: codecs: aw88166: remove async start

Codec drivers are not supposed to do anything like this. The result was
that the first second or so of playback was essentially inaudible, and
very short alert sounds could be missed entirely. Let's not do this.

Adapted for aw88166 from Val Packett's aw88261 series.
---
 sound/soc/codecs/aw88166.c | 34 ++++------------------------------
 sound/soc/codecs/aw88166.h |  5 -----
 2 files changed, 4 insertions(+), 35 deletions(-)

diff --git a/sound/soc/codecs/aw88166.c b/sound/soc/codecs/aw88166.c
index af0b00f..326ffba 100644
--- a/sound/soc/codecs/aw88166.c
+++ b/sound/soc/codecs/aw88166.c
@@ -23,7 +23,6 @@ struct aw88166 {
 	struct aw_device *aw_pa;
 	struct mutex lock;
 	struct gpio_desc *reset_gpio;
-	struct delayed_work start_work;
 	struct regmap *regmap;
 	struct aw_container *aw_cfg;
 
@@ -1135,17 +1134,7 @@ static void aw88166_start_pa(struct aw88166 *aw88166)
 		dev_err(aw88166->aw_pa->dev, "start failure (%d)\n", ret);
 }
 
-static void aw88166_startup_work(struct work_struct *work)
-{
-	struct aw88166 *aw88166 =
-		container_of(work, struct aw88166, start_work.work);
-
-	mutex_lock(&aw88166->lock);
-	aw88166_start_pa(aw88166);
-	mutex_unlock(&aw88166->lock);
-}
-
-static void aw88166_start(struct aw88166 *aw88166, bool sync_start)
+static void aw88166_start(struct aw88166 *aw88166)
 {
 	int ret;
 
@@ -1161,12 +1150,7 @@ static void aw88166_start(struct aw88166 *aw88166, bool sync_start)
 		return;
 	}
 
-	if (sync_start == AW88166_SYNC_START)
-		aw88166_start_pa(aw88166);
-	else
-		queue_delayed_work(system_dfl_wq,
-			&aw88166->start_work,
-			AW88166_START_WORK_DELAY_MS);
+	aw88166_start_pa(aw88166);
 }
 
 static int aw_dev_check_sysint(struct aw_device *aw_dev)
@@ -1453,7 +1437,7 @@ static int aw88166_profile_set(struct snd_kcontrol *kcontrol,
 
 	if (aw88166->aw_pa->status) {
 		aw88166_stop(aw88166->aw_pa);
-		aw88166_start(aw88166, AW88166_SYNC_START);
+		aw88166_start(aw88166);
 	}
 
 	mutex_unlock(&aw88166->lock);
@@ -1632,7 +1616,7 @@ static int aw88166_playback_event(struct snd_soc_dapm_widget *w,
 	mutex_lock(&aw88166->lock);
 	switch (event) {
 	case SND_SOC_DAPM_PRE_PMU:
-		aw88166_start(aw88166, AW88166_ASYNC_START);
+		aw88166_start(aw88166);
 		break;
 	case SND_SOC_DAPM_POST_PMD:
 		aw88166_stop(aw88166->aw_pa);
@@ -1667,8 +1651,6 @@ static int aw88166_codec_probe(struct snd_soc_component *component)
 	struct aw88166 *aw88166 = snd_soc_component_get_drvdata(component);
 	int ret;
 
-	INIT_DELAYED_WORK(&aw88166->start_work, aw88166_startup_work);
-
 	ret = aw88166_request_firmware_file(aw88166);
 	if (ret)
 		dev_err(aw88166->aw_pa->dev, "%s failed\n", __func__);
@@ -1676,16 +1658,8 @@ static int aw88166_codec_probe(struct snd_soc_component *component)
 	return ret;
 }
 
-static void aw88166_codec_remove(struct snd_soc_component *aw_codec)
-{
-	struct aw88166 *aw88166 = snd_soc_component_get_drvdata(aw_codec);
-
-	cancel_delayed_work_sync(&aw88166->start_work);
-}
-
 static const struct snd_soc_component_driver soc_codec_dev_aw88166 = {
 	.probe = aw88166_codec_probe,
-	.remove = aw88166_codec_remove,
 	.dapm_widgets = aw88166_dapm_widgets,
 	.num_dapm_widgets = ARRAY_SIZE(aw88166_dapm_widgets),
 	.dapm_routes = aw88166_audio_map,
diff --git a/sound/soc/codecs/aw88166.h b/sound/soc/codecs/aw88166.h
index 846413d..adb8314 100644
--- a/sound/soc/codecs/aw88166.h
+++ b/sound/soc/codecs/aw88166.h
@@ -556,7 +556,6 @@
 #define AW88166_CRC_FW_BASE_ADDR		(0x4C0)
 #define AW88166_DEV_SYSST_CHECK_MAX		(10)
 #define AW88166_START_RETRIES			(5)
-#define AW88166_START_WORK_DELAY_MS		(0)
 #define AW88166_CHIP_ID			(0x2066)
 #define AW88166_I2C_NAME			"aw88166"
 #define AW88166_ACF_FILE			"aw88166_acf.bin"
@@ -621,10 +620,6 @@ enum AW88166_DEV_DSP_CFG {
 	AW88166_DEV_DSP_BYPASS = 1,
 };
 
-enum {
-	AW88166_SYNC_START = 0,
-	AW88166_ASYNC_START,
-};
 
 enum {
 	AW88166_RECORD_SEC_DATA = 0,
-- 
2.55.0

